File: No_2328/Number of Increasing Paths in a Grid.cpp
class Solution {
public:
    int MOD = 1e9 + 7;
    int countPaths(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, -1));
        
        int ret = 0;
        for (int i = 0; i < m ; i++) {
            for (int j = 0; j < n; j++) {
                ret = (ret + dfs(grid, dp, i, j)) % MOD;
            }
        }
        
        return ret;
    }
    
    int dfs(vector<vector<int>>& grid, vector<vector<int>>& dp, int i, int j) {
        if (dp[i][j] != -1)
            return dp[i][j];
        
        int pathCount = 0;
        if (i - 1 >= 0 && grid[i-1][j] < grid[i][j])
            pathCount = (pathCount + dfs(grid, dp, i-1, j)) % MOD;
        if (i + 1 < grid.size() && grid[i+1][j] < grid[i][j])
            pathCount = (pathCount + dfs(grid, dp, i+1, j)) % MOD;
        if (j - 1 >= 0 && grid[i][j-1] < grid[i][j])
            pathCount = (pathCount + dfs(grid, dp, i, j-1)) % MOD;
        if (j + 1 < grid[0].size() && grid[i][j + 1] < grid[i][j])
            pathCount = (pathCount + dfs(grid, dp, i, j+1)) % MOD;
        
        dp[i][j] = pathCount + 1;
        
        return dp[i][j];
    }
};

File: No_2328/Number of Increasing Paths in a Grid.py
class Solution:
    def countPaths(self, grid: List[List[int]]) -> int:
        def dfs(grid, dp, i, j):
            if (i, j) in dp:
                return dp[(i, j)]
            
            path_count = 0
            if i - 1 >= 0 and grid[i-1][j] < grid[i][j]:
                path_count += dfs(grid, dp, i-1, j)
            if i + 1 < len(grid) and grid[i+1][j] < grid[i][j]:
                path_count += dfs(grid, dp, i+1, j)
            if j - 1 >= 0 and grid[i][j-1] < grid[i][j]:
                path_count += dfs(grid, dp, i, j-1)
            if j + 1 < len(grid[0]) and grid[i][j+1] < grid[i][j]:
                path_count += dfs(grid, dp, i, j+1)
        
            
            dp[(i,j)] = path_count + 1
            return dp[(i,j)]
            
        ret = 0
        dp = {}
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                ret += dfs(grid, dp, i, j)
                
        return ret % (10**9 + 7)